File: api/src/main/scala/com/github/pflooky/datacaterer/api/PlanBuilder.scala
package com.github.pflooky.datacaterer.api

import com.github.pflooky.datacaterer.api.connection.ConnectionTaskBuilder
import com.github.pflooky.datacaterer.api.converter.Converters.toScalaList
import com.github.pflooky.datacaterer.api.model.{ForeignKeyRelation, Plan, SinkOptions}
import com.softwaremill.quicklens.ModifyPimp

import scala.annotation.varargs

case class PlanBuilder(plan: Plan = Plan(), tasks: List[TasksBuilder] = List()) {
  def this() = this(Plan(), List())

  def name(name: String): PlanBuilder =
    this.modify(_.plan.name).setTo(name)

  def description(desc: String): PlanBuilder =
    this.modify(_.plan.description).setTo(desc)

  def taskSummaries(taskSummaries: TaskSummaryBuilder*): PlanBuilder = {
    val tasksToAdd = taskSummaries.filter(_.task.isDefined)
      .map(x => TasksBuilder(List(x.task.get), x.taskSummary.dataSourceName))
      .toList
    this.modify(_.plan.tasks)(_ ++ taskSummaries.map(_.taskSummary))
      .modify(_.tasks)(_ ++ tasksToAdd)
  }

  def sinkOptions(sinkOptionsBuilder: SinkOptionsBuilder): PlanBuilder =
    this.modify(_.plan.sinkOptions).setTo(Some(sinkOptionsBuilder.sinkOptions))

  def sinkOptions(sinkOptions: SinkOptions): PlanBuilder =
    this.modify(_.plan.sinkOptions).setTo(Some(sinkOptions))

  def seed(seed: Long): PlanBuilder =
    this.modify(_.plan.sinkOptions).setTo(Some(getSinkOpt.seed(seed).sinkOptions))

  def locale(locale: String): PlanBuilder =
    this.modify(_.plan.sinkOptions).setTo(Some(getSinkOpt.locale(locale).sinkOptions))

  @varargs def addForeignKeyRelationship(foreignKey: ForeignKeyRelation, relations: ForeignKeyRelation*): PlanBuilder =
    this.modify(_.plan.sinkOptions).setTo(Some(getSinkOpt.foreignKey(foreignKey, relations.toList).sinkOptions))

  def addForeignKeyRelationship(connectionTaskBuilder: ConnectionTaskBuilder[_], columns: List[String],
                                relations: List[(ConnectionTaskBuilder[_], List[String])]): PlanBuilder = {
    val baseRelation = toForeignKeyRelation(connectionTaskBuilder, columns)
    val otherRelations = relations.map(r => toForeignKeyRelation(r._1, r._2))
    addForeignKeyRelationship(baseRelation, otherRelations: _*)
  }

  def addForeignKeyRelationship(connectionTaskBuilder: ConnectionTaskBuilder[_], columns: java.util.List[String],
                                relations: java.util.List[java.util.Map.Entry[ConnectionTaskBuilder[_], java.util.List[String]]]): PlanBuilder = {
    val scalaListRelations = toScalaList(relations)
    val mappedRelations = scalaListRelations.map(r => (r.getKey, toScalaList(r.getValue)))
    addForeignKeyRelationship(connectionTaskBuilder, toScalaList(columns), mappedRelations)
  }

  def addForeignKeyRelationship(connectionTaskBuilder: ConnectionTaskBuilder[_], column: String,
                                relations: List[(ConnectionTaskBuilder[_], String)]): PlanBuilder =
    addForeignKeyRelationship(connectionTaskBuilder, List(column), relations.map(r => (r._1, List(r._2))))

  def addForeignKeyRelationship(connectionTaskBuilder: ConnectionTaskBuilder[_], column: String,
                                relations: java.util.List[java.util.Map.Entry[ConnectionTaskBuilder[_], String]]): PlanBuilder = {
    val scalaListRelations = toScalaList(relations)
    val mappedRelations = scalaListRelations.map(r => (r.getKey, List(r.getValue)))
    addForeignKeyRelationship(connectionTaskBuilder, List(column), mappedRelations)
  }

  def addForeignKeyRelationships(connectionTaskBuilder: ConnectionTaskBuilder[_], columns: List[String],
                                 relations: List[ForeignKeyRelation]): PlanBuilder = {
    val baseRelation = toForeignKeyRelation(connectionTaskBuilder, columns)
    addForeignKeyRelationship(baseRelation, relations: _*)
  }

  def addForeignKeyRelationship(foreignKey: ForeignKeyRelation,
                                relations: List[(ConnectionTaskBuilder[_], List[String])]): PlanBuilder =
    addForeignKeyRelationship(foreignKey, relations.map(r => toForeignKeyRelation(r._1, r._2)): _*)

  private def toForeignKeyRelation(connectionTaskBuilder: ConnectionTaskBuilder[_], columns: List[String]) = {
    val dataSource = connectionTaskBuilder.connectionConfigWithTaskBuilder.dataSourceName
    val colNames = columns.mkString(",")
    connectionTaskBuilder.step match {
      case Some(value) =>
        val fields = value.step.schema.fields.getOrElse(List())
        val hasColumns = columns.forall(c => fields.exists(_.name == c))
        if (!hasColumns) {
          throw new RuntimeException(s"Column name defined in foreign key relationship does not exist, data-source=$dataSource, column-name=$colNames")
        }
        ForeignKeyRelation(dataSource, value.step.name, columns)
      case None =>
        throw new RuntimeException(s"Not schema defined for data source. Cannot create foreign key relationship, data-source=$dataSource, column-name=$colNames")
    }
  }

  private def getSinkOpt: SinkOptionsBuilder = {
    plan.sinkOptions match {
      case Some(value) => SinkOptionsBuilder(value)
      case None => SinkOptionsBuilder()
    }
  }
}

File: api-java/src/main/java/com/github/pflooky/datacaterer/api/java/PlanBuilder.java
package com.github.pflooky.datacaterer.api.java;

import com.github.pflooky.datacaterer.api.java.model.SinkOptionsBuilder;
import com.github.pflooky.datacaterer.api.java.model.TaskSummaryBuilder;
import com.github.pflooky.datacaterer.api.model.ForeignKeyRelation;
import com.github.pflooky.datacaterer.api.model.Plan;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaList;
import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaOption;
import static com.github.pflooky.datacaterer.api.converter.Converters.toScalaSeq;

public final class PlanBuilder {
    private final com.github.pflooky.datacaterer.api.PlanBuilder scalaDef;

    public PlanBuilder(com.github.pflooky.datacaterer.api.PlanBuilder scalaDef) {
        this.scalaDef = scalaDef;
    }

    public PlanBuilder() {
        this.scalaDef = new com.github.pflooky.datacaterer.api.PlanBuilder(
                new Plan(
                        "Default plan",
                        "Data generation plan",
                        toScalaList(Collections.emptyList()),
                        toScalaOption(Optional.empty()),
                        toScalaList(Collections.emptyList())
                ), toScalaList(Collections.emptyList())
        );
    }

    public com.github.pflooky.datacaterer.api.PlanBuilder plan() {
        return scalaDef;
    }

    public PlanBuilder name(String name) {
        return new PlanBuilder(scalaDef.name(name));
    }

    public PlanBuilder description(String description) {
        return new PlanBuilder(scalaDef.description(description));
    }

    public PlanBuilder taskSummaries(TaskSummaryBuilder... taskSummaryBuilders) {
        return new PlanBuilder(scalaDef.taskSummaries(
                toScalaSeq(Arrays.stream(taskSummaryBuilders).map(TaskSummaryBuilder::taskSummary).collect(Collectors.toList()))
        ));
    }

    public PlanBuilder sinkOptions(SinkOptionsBuilder sinkOptionsBuilder) {
        return new PlanBuilder(scalaDef.sinkOptions(sinkOptionsBuilder.sinkOptions()));
    }

    public PlanBuilder seed(long seed) {
        return new PlanBuilder(scalaDef.seed(seed));
    }

    public PlanBuilder locale(String locale) {
        return new PlanBuilder(scalaDef.locale(locale));
    }

    public PlanBuilder addForeignKeyRelationship(ForeignKeyRelation foreignKey, ForeignKeyRelation relation, ForeignKeyRelation... relations) {
        return new PlanBuilder(scalaDef.addForeignKeyRelationship(foreignKey, relation, toScalaSeq(Arrays.asList(relations))));
    }

    public PlanBuilder addForeignKeyRelationship(ForeignKeyRelation foreignKey, List<ForeignKeyRelation> relations) {
        return new PlanBuilder(scalaDef.addForeignKeyRelationship(foreignKey, toScalaList(relations)));
    }
}
